vent/coordinator/coordinator.py

```diff
@@ -72,127 +121,123 @@ def __init__(self, sim_mode=False):
-
 
 class CoordinatorRemote(CoordinatorBase):
     def __init__(self, sim_mode=False):
         super().__init__(sim_mode=sim_mode)
-        # TODO: pass max_heartbeat_interval
-        self.process_manager = ProcessManager()
-        self.rpc = IPC()
-        raise NotImplementedError
+        # TODO: according to documentation, pass max_heartbeat_interval
+        self.ipc = IPC(listen=True)
+        self.process_manager = ProcessManager(sim_mode, self.ipc.addr, self.ipc.port)
+        # TODO: make sure the ipc connection is setup. There should be a clever method
+        time.sleep(1)
+        self.receive_thread = threading.Thread(target=self.__receive_loop, daemon=True)
+        self.receive_thread.start()
+        self.receive_thread_id = self.receive_thread.ident
+        self.send_thread = threading.Thread(target=self.__send_loop, daemon=True)
+        self.send_thread.start()
+        self.send_thread_id = self.send_thread.ident
+        self.start()
+
+    def start(self):
+        # send start msg
+        self.ipc.send_msg(IPCMessage(IPCCommand.START))
+        super().start()
+
+    def stop(self):
+        # send stop msg
+        super().stop()
+        self.ipc.send_msg(IPCMessage(IPCCommand.STOP))
+
+    def __receive_loop(self):
+        """
+        The key to prevent deadlock is this thread never blocking
+        """
+        while self._is_running.wait():
+            # TODO: we need have better logging!
+            # print('parent in receiving loop')
+            msg = self.ipc.recv_msg()
+            # print(f'parent process received: {msg.command}')
+            if msg.command == IPCCommand.GET_SENSORS:
+                sensor_values = msg.args
+                # print(f'parent process read sensor: {sensor_values}')
+                with self.lock:
+                    self.COPY_sensor_values = sensor_values
+                # TODO: set timestamp
+            elif msg.command == IPCCommand.GET_CONTROL:
+                if msg.args is not None:
+                    control_setting = msg.args
+                    name = control_setting.name
+                    with self.lock:
+                        self.COPY_control_settings[name] = control_setting
+                else:
+                    # assume wait some time will sync the control setting
+                    pass
+            # TODO: if msg is alarm
+            else:
+                raise NotImplementedError(f'Error: {msg.command} not implemented')
+
+    def __send_loop(self):
+        while self._is_running.wait():
+            # don't send GET_SENSOR command because assuming it will send periodically
+            # print('parent in sending loop')
+            for name in values.controllable_values:
+                with self.lock:
+                    not_in_control_settings = name not in self.COPY_control_settings
+                if not_in_control_settings:
+                    self.ipc.send_msg(IPCMessage(IPCCommand.GET_CONTROL, name))
```

Comment:
there's no guarantee this completes by the time you go to check on a change in line 226, so this probably will result in commands being set multiple times. one of the benefits of having message handling being in the same method.

Reply:
I propose to handle this in next pull request.

Reply:
sure np